### PR TITLE
Cythonize CuPy

### PR DESCRIPTION
**Pull request by okuta**
Python code is very slow.

I cythonize core feature of CuPy in this PR.
- cupy.ndarray
- cupy.cuda
- cupy.carray
- cupy.elementwise
- cupy.reduction
- And other if necessary

This PR relates #499, #399 and #333.

### CODE REVIEW

File: cupy/cuda/cupy_cuda.h [hunk 1]
```diff
@@ -0,0 +1,517 @@
+// This file is a stub header file of cuda for Read the Docs.
+
+#ifndef INCLUDE_GUARD_CUPY_CUDA_H
+#define INCLUDE_GUARD_CUPY_CUDA_H
+
+#ifndef CUPY_NO_CUDA
+#include <cuda.h>
+#include <cuda_runtime.h>
+#include <cublas_v2.h>
```
Comment (unnonouno) [inline cupy/cuda/cupy_cuda.h]: Alphabet order?

**Comment by okuta** [comment]
Performance test results.

mnist
```
v1.4.1   ./train_mnist.py --gpu 0  66.04s user 2.18s system 99% cpu 1:08.78 total
v1.2     ./train_mnist.py --gpu 0  44.01s user 3.54s system 99% cpu 47.773 total
this PR  ./train_mnist.py --gpu 0  40.45s user 1.77s system 98% cpu 42.791 total
````

ptb
```
v1.2     iter 10000 training perplexity: 1008.17 (204.81 iters/sec)
v1.4.1   iter 10000 training perplexity: 984.55 (130.01 iters/sec)
this PR  iter 10000 training perplexity: 952.01 (165.56 iters/sec)
```

**Comment by unnonouno** [comment]
It looks good. Great!!

**Comment by unnonouno** [comment]
We'll fix trivial problems after merge such as code styles in pyx files.

**Pull request merged**